clang/docs/LanguageExtensions.rst

```diff
@@ -96,6 +100,47 @@ the ``<cmath>`` header file to conditionally make a function constexpr whenever
 the constant evaluation of the corresponding builtin (for example,
 ``std::fmax`` calls ``__builtin_fmax``) is supported in Clang.
 
+``__has_target_builtin``
+------------------------
+
+This function-like macro takes a single identifier argument that is the name of
+a builtin function, a builtin pseudo-function (taking one or more type
+arguments), or a builtin template.
+It evaluates to 1 if the builtin is supported on the current target or 0 if not.
+
+``__has_builtin`` and ``__has_target_builtin`` behave identically for normal C++ compilations.
```

Comment:
So this means users should use `__has_target_builtin` in place of `__has_builtin` so that their code will correctly check for the builtin on host or device?

Reply:
if the user's goal is to check that the builtin can be codegen'd on the current target being compiled, then yes they should use `__has_target_builtin`. if they want to confirm it can be parsed but not necessarily codegen'd, then they should use `__has_builtin`. if that's unclear let me know and i can try to improve it

Reply:
> if the user's goal is to check that the builtin can be codegen'd on the current target being compiled, then yes they should use __has_target_builtin. if they want to confirm it can be parsed but not necessarily codegen'd, then they should use __has_builtin. if that's unclear let me know and i can try to improve the doc

Users don't typically think in terms of "parsed" and "codegenned", but more "works" and "doesn't work", which I think means "codegenned" in general. The line we're drawing here is pretty subtle, so perhaps more real world examples would help; it's hard to understand why a builtin can be parsed but cannot be used.

Reply:
yeah that's a fair point, so it sounds like you'd prefer an example of when you codegen doesn't work even though you are checking with `has_builtin`, and for that i could show the motiving example for this which is something like

```
void foo() {
#if __has_builtin(__builtin_ia32_pause)
__builtin_ia32_pause()
#else
abort()
#endif
}
```
and if the current target is an offloading target (amdgpu/nvptx/spirv) and the aux target is x86, we will get a error saying it can't be codegen'd

would that kind of example address your concern? thx

Reply:
Yeah, I think it would, along with the prose explaining why that's not a bug but is actually by design for `__has_builtin`. I think users would reasonably look at that and say it should not be an error, so it'd be nice to make sure they understand the intent.

Reply:
sure, sounds good. 

@yxsamliu @Artem-B would implementing Aaron's suggestion address your concerns as well?

Reply:
+1 to including a concrete example with an explanation. The fact that a builtin may be both visible but not usable will be a surprise for a C++ user.

Reply:
thanks, ill add this

Reply:
added in the latest commit, please take a look and let me know if it's clear